### PR TITLE
Add RoBERTa base model to applications

### PR DESCRIPTION
**Pull request by mrwyattii**
Adds the RoBERTa base model (based on [HuggingFace implementation](https://huggingface.co/docs/transformers/v4.15.0/en/model_doc/roberta#transformers.RobertaModel)). Also added a script for downloading model configuration and pretrained weights from huggingface. An example of how to run the model with LBANN is provided in `main.py`

Includes changes from #2019

### CODE REVIEW

File: applications/nlp/RoBERTa/main.py [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+    # return 0.5 * x * (1 + tanh(sqrt(pi / 2) * (x + 0.044715 * x ** 3)))
+    # Based on: https://github.com/pytorch/pytorch/issues/20464#issuecomment-492339005
+    ones = _ValueTensor(x_shape, 1)
+    twos = _ValueTensor(x_shape, 2)
+    sqrt_pi_over_2 = _ValueTensor(x_shape, math.sqrt(math.pi / 2))
+    b_coef = _ValueTensor(x_shape, 0.044715)
+    x_cubed = lbann.Multiply(lbann.Multiply(lbann.Identity(x), x), x)
+    inner_tanh_x_comp = lbann.Add(x, lbann.Multiply(b_coef, x_cubed))
+    tanh_x = lbann.Tanh(lbann.Multiply(sqrt_pi_over_2, inner_tanh_x_comp))
+    return lbann.Divide(lbann.Multiply(x, lbann.Add(ones, tanh_x)), twos)
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:597]:
```suggestion
def _gelu(x, x_shape):
    # return 0.5 * x * (1 + tanh(sqrt(pi / 2) * (x + 0.044715 * x ** 3)))
    # Based on: https://github.com/pytorch/pytorch/issues/20464#issuecomment-492339005
    sqrt_pi_over_2 = math.sqrt(math.pi / 2)
    b_coef = 0.044715
    x_cubed = lbann.Multiply(lbann.Multiply(lbann.Identity(x), x), x)
    inner_tanh_x_comp = lbann.Add(x, lbann.Scale(x_cubed, constant=b_coef))
    tanh_x = lbann.Tanh(lbann.Scale(inner_tanh_x_comp, constant=sqrt_pi_over_2))
    return lbann.Scale(lbann.Multiply(x, lbann.AddConstant(tanh_x, constant=1)), constant=0.5)
```

File: applications/nlp/RoBERTa/main.py [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+            tmp_weights = lbann.Weights(
+                initializer=lbann.ValueInitializer(
+                    values=lbann.util.str_list(np.nditer(attention_mask))
+                ),
+                optimizer=lbann.NoOptimizer(),
+            )
+            attention_mask = lbann.WeightsLayer(
+                dims=lbann.util.str_list(self.input_shape),
+                weights=tmp_weights,
+            )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:770]:
```suggestion
            attention_mask = lbann.Constant(
                value=1,
                num_neurons=lbann.utils.str_list(self.input_shape),
            )
```

File: applications/nlp/RoBERTa/main.py [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+            tmp_weights = lbann.Weights(
+                initializer=lbann.ValueInitializer(
+                    values=lbann.util.str_list(np.nditer(token_type_ids))
+                ),
+                optimizer=lbann.NoOptimizer(),
+            )
+            token_type_ids = lbann.WeightsLayer(
+                dims=lbann.util.str_list(self.input_shape),
+                weights=tmp_weights,
+            )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:783]:
```suggestion
            token_type_ids = lbann.Constant(
                value=0,
                num_neurons=lbann.util.str_list(self.input_shape),
            )
```

File: applications/nlp/RoBERTa/main.py [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+    tmp_weights = lbann.Weights(
+        initializer=lbann.ValueInitializer(
+            values=lbann.util.str_list(np.nditer(padding_idx))
+        ),
+        optimizer=lbann.NoOptimizer(),
+    )
+    padding_idx = lbann.WeightsLayer(
+        dims=lbann.util.str_list(input_shape),
+        weights=tmp_weights,
+    )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:877]:
```suggestion
    padding_idx = lbann.Constant(
        value=padding_idx,
        dims=lbann.util.str_list(input_shape),
    )
```

File: applications/nlp/RoBERTa/main.py [hunk 5]
```diff
@@ -0,0 +1,992 @@
...
+    tmp_weights = lbann.Weights(
+        initializer=lbann.ValueInitializer(
+            values=lbann.util.str_list(np.nditer(past_key_values_length))
+        ),
+        optimizer=lbann.NoOptimizer(),
+    )
+    past_key_values_length = lbann.WeightsLayer(
+        dims=lbann.util.str_list(input_shape),
+        weights=tmp_weights,
+    )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:894]:
```suggestion
    past_key_values_length = lbann.Constant(
        value=past_key_values_length,
        num_neurons=lbann.util.str_list(input_shape),
    )
```

File: applications/nlp/RoBERTa/main.py [hunk 6]
```diff
@@ -0,0 +1,992 @@
...
+    x = lbann.Reshape(x, dims=str_list([size]))
+    y = lbann.Gather(x, inds)
+    y = lbann.Reshape(y, dims=str_list(list(new_dims)))
+    return y, tuple(new_dims)
+
+
+def _cumsum(x, dims, axis=0):
+    global _cumsum_cache
+
+    if len(dims) > 2:
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:934]:
```suggestion
    if len(dims) != 2:
```
Handling `len(dims)==1` will just require some extra reshaping, if it's needed.

File: applications/nlp/RoBERTa/main.py [hunk 7]
```diff
@@ -0,0 +1,992 @@
...
+
+    # Apply cumsum
+    tril_ones = _cumsum_cache[shape]
+    if axis == 0:
+        x = lbann.MatMul(tril_ones, x)
+        return x
+    if axis == 1:
+        x, _ = _permute(x, dims)
+        x = lbann.MatMul(tril_ones, x)
+        x, _ = _permute(x, dims[::-1])
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:958]:
```suggestion
        x = lbann.MatMul(x, tril_ones, transpose_b=True)
```

**Review by timmoon10 — COMMENTED** [review]
Overall looks good. In addition to the recommendations, I'd recommend throwing exceptions with:
```
sed 's/sys\.exit/raise RuntimeError/'
```

File: applications/nlp/RoBERTa/main.py [hunk 8]
```diff
@@ -0,0 +1,760 @@
...
+    # return 0.5 * x * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * x ** 3)))
+    # Based on: https://github.com/huggingface/transformers/blob/38a716cd41f22f6a7d5ff3dc081903090198803a/examples/research_projects/bertabs/modeling_bertabs.py#L658
+    sqrt_2_over_pi = math.sqrt(2 / math.pi)
+    b_coef = 0.044715
+    x_cubed = lbann.Multiply(lbann.Multiply(lbann.Identity(x), x), x)
+    inner_tanh_x_comp = lbann.Add(x, lbann.Scale(x_cubed, constant=b_coef))
+    tanh_x = lbann.Tanh(lbann.Scale(inner_tanh_x_comp, constant=sqrt_2_over_pi))
+    return lbann.Scale(
+        lbann.Multiply(x, lbann.AddConstant(tanh_x, constant=1)), constant=0.5
+    )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/main.py:543]: Are you missing a multiplication by `x`?

**Review by bvanessen — APPROVED** [review]

**Comment by bvanessen** [comment]
@mrwyattii I am assuming that we should merge #2019 first

File: applications/nlp/RoBERTa/roberta.py [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+        input_ids = lbann.Reshape(input_ids, dims=str_list(self.input_shape))
+        embedding_output = self.embeddings(
+            input_ids=input_ids,
+            position_ids=position_ids,
+            token_type_ids=token_type_ids,
+            inputs_embeds=inputs_embeds,
+        )
+        embedding_output = lbann.Reshape(
+            embedding_output, dims=str_list(self.input_shape + (768,))
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/roberta.py:586]:
```suggestion
            embedding_output, dims=str_list(self.input_shape + (self.hidden_size,))
```

**Review by timmoon10 — APPROVED** [review]
Overall LGTM. Your layer-by-layer comparisons with PyTorch gave me a lot of confidence.

File: applications/nlp/RoBERTa/roberta.py [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+):
+    padding_idx = lbann.Constant(value=padding_idx, num_neurons=str_list(input_shape))
+    mask = lbann.NotEqual(input_ids, padding_idx)
+    incremental_indices = lbann.modules.Cumsum(mask, input_shape, axis=1)
+    past_key_values_length = lbann.Constant(
+        value=past_key_values_length, num_neurons=str_list(input_shape)
+    )
+    incremental_indices = lbann.Add(incremental_indices, past_key_values_length)
+    incremental_indices = lbann.Multiply(incremental_indices, mask)
+    incremental_indices = lbann.Add(incremental_indices, padding_idx)
```
[the remark this answers is not in the document]
Comment (mrwyattii) [reply applications/nlp/RoBERTa/roberta.py:33]: cleaned up this function in most recent commit.

File: applications/nlp/RoBERTa/roberta.py [hunk 11]
```diff
@@ -161,6 +162,22 @@ def forward(
...
+        )
+        position_ids = lbann.WeightsLayer(
+            weights=lbann.Weights(
+                initializer=lbann.ValueInitializer(values=str_list(position_ids))
+            ),
+            dims=str_list([sequence_length]),
+        )
+        position_ids = lbann.Tessellate(
+            position_ids, dims=str_list(self.input_shape[:-1])
+        )
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/roberta.py:178]: Why do we need to tessellate to match the embedding vector size? If we are throwing these indices into an embedding layer, shouldn't that bring us up to the right shape?

File: applications/nlp/RoBERTa/roberta.py [hunk 12]
```diff
@@ -161,6 +162,22 @@ def forward(
...
 
+    def create_position_ids_from_inputs_embeds(self, input_embeds):
+        sequence_length = self.input_shape[1]
+        position_ids = range(
+            self.padding_idx + 1, sequence_length + self.padding_idx + 1
+        )
+        position_ids = lbann.WeightsLayer(
+            weights=lbann.Weights(
+                initializer=lbann.ValueInitializer(values=str_list(position_ids))
+            ),
```
Comment (timmoon10) [inline applications/nlp/RoBERTa/roberta.py:173]: Just to be on the safe side:
```suggestion
            weights=lbann.Weights(
                initializer=lbann.ValueInitializer(values=str_list(position_ids))
                optimizer=lbann.NoOptimizer(),
            ),
```

File: applications/nlp/RoBERTa/roberta.py (hunk 11 above)
Comment (timmoon10) [inline applications/nlp/RoBERTa/roberta.py:178]: You may need to reshape `position_ids` with dims of size 1 to make sure it has the right number of dims.

Comment (mrwyattii) [reply applications/nlp/RoBERTa/roberta.py:180]: Fixed with most recent commit

File: applications/nlp/RoBERTa/roberta.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrwyattii) [reply applications/nlp/RoBERTa/roberta.py:33]: positions begin at `padding_idx + 1`. See implementation here: https://github.com/huggingface/transformers/blob/baf1ebe9f033b0d8cd1ca095ba68f6789ff3d505/src/transformers/models/roberta/modeling_roberta.py#L1565

**Pull request merged**